File: golang/tests/pkg/integration/ws_stream_cancel_integration_test.go
package integration_test

import (
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/agentbay/testutil"
)

func TestWsStreamCancelE2E(t *testing.T) {
	apiKey := testutil.GetTestAPIKey(t)
	ab, err := agentbay.NewAgentBay(apiKey)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	imageID := "imgc-0ab5taki2khozz0p8"
	if v := os.Getenv("AGENTBAY_WS_IMAGE_ID"); v != "" {
		imageID = v
	}
	params := agentbay.NewCreateSessionParams().WithImageId(imageID)
	sessionResult, err := ab.Create(params)
	if err != nil {
		t.Fatalf("Error creating session: %v", err)
	}
	session := sessionResult.Session
	defer func() {
		_, _ = ab.Delete(session)
	}()

	wsClient, err := session.GetStreamingWsClient()
	if err != nil {
		t.Fatalf("Error getting WS client: %v", err)
	}
	defer func() {
		_ = wsClient.Close()
	}()

	target := "wuying_codespace"
	for _, tool := range session.McpTools {
		if tool.GetName() == "run_code" && tool.GetServer() != "" {
			target = tool.GetServer()
			break
		}
	}

	events := []map[string]interface{}{}
	ends := []map[string]interface{}{}
	errors := []error{}

	onEvent := func(_invocationID string, data map[string]interface{}) {
		events = append(events, data)
	}
	onEnd := func(_invocationID string, data map[string]interface{}) {
		ends = append(ends, data)
	}
	onError := func(_invocationID string, err error) {
		errors = append(errors, err)
	}

	handle, err := wsClient.CallStream(target, map[string]interface{}{
		"method": "run_code",
		"mode":   "stream",
		"params": map[string]interface{}{
			"language": "python",
			"timeoutS": 60,
			"code": "import time\n" +
				"print(0)\n" +
				"time.sleep(10)\n" +
				"print(1)\n",
		},
	}, onEvent, onEnd, onError)
	if err != nil {
		t.Fatalf("call stream error: %v", err)
	}

	time.Sleep(500 * time.Millisecond)
	if err := handle.Cancel(); err != nil {
		t.Fatalf("cancel error: %v", err)
	}

	type waitRes struct {
		data map[string]interface{}
		err  error
	}
	resCh := make(chan waitRes, 1)
	go func() {
		d, e := handle.WaitEnd()
		resCh <- waitRes{data: d, err: e}
	}()

	select {
	case r := <-resCh:
		if r.err == nil {
			t.Fatalf("expected WsCancelledError, got nil err data=%v", r.data)
		}
		if _, ok := r.err.(*agentbay.WsCancelledError); !ok {
			t.Fatalf("expected WsCancelledError, got %T: %v", r.err, r.err)
		}
	case <-time.After(2 * time.Second):
		t.Fatalf("timeout waiting for WaitEnd after cancel")
	}

	if len(ends) != 0 {
		t.Fatalf("unexpected onEnd after cancel: ends=%v events=%v errors=%v", ends, events, errors)
	}
	if len(errors) != 1 {
		t.Fatalf("expected exactly 1 onError, got=%v", errors)
	}
	if _, ok := errors[0].(*agentbay.WsCancelledError); !ok {
		t.Fatalf("expected onError WsCancelledError, got %T: %v", errors[0], errors[0])
	}
}

File: typescript/tests/integration/ws-stream-cancel.integration.test.ts
import { AgentBay, Session, WsCancelledError } from "../../src";
import { log, logError } from "../../src/utils/logger";

describe("WS stream cancel Integration Test", () => {
  let agentBay: AgentBay | null = null;
  let session: Session | null = null;

  jest.setTimeout(120000);

  beforeAll(async () => {
    const apiKey = process.env.AGENTBAY_API_KEY;
    if (!apiKey) {
      log(
        "AGENTBAY_API_KEY is not set; skipping WS stream cancel integration test"
      );
      return;
    }

    agentBay = new AgentBay({ apiKey });
    const imageId =
      process.env.AGENTBAY_WS_IMAGE_ID || "imgc-0ab5taki2khozz0p8";
    const created = await agentBay.create({ imageId });
    if (!created.success || !created.session) {
      throw new Error(`Failed to create session: ${created.errorMessage}`);
    }
    const s = created.session;
    session = s;
    if (!s.wsUrl) {
      throw new Error("wsUrl is empty in CreateMcpSession response");
    }
  });

  afterAll(async () => {
    if (session) {
      try {
        await session.delete();
      } catch (e) {
        logError("Failed to delete session:", e);
      }
    }
  });

  test("cancel should reject waitEnd with WsCancelledError", async () => {
    if (!session) return;

    const wsClient = await session.getWsClient();
    let target = "wuying_codespace";
    for (const tool of session.mcpTools || []) {
      try {
        if (tool && tool.name === "run_code" && tool.server) {
          target = tool.server;
          break;
        }
      } catch (_e) {
        continue;
      }
    }

    const events: Record<string, any>[] = [];
    const ends: Record<string, any>[] = [];
    const errors: Error[] = [];

    const handle = await wsClient.callStream({
      target,
      data: {
        method: "run_code",
        mode: "stream",
        params: {
          language: "python",
          timeoutS: 60,
          code:
            "import time\n" +
            "print(0, flush=True)\n" +
            "time.sleep(10)\n" +
            "print(1, flush=True)\n",
        },
      },
      onEvent: (_invocationId, data) => {
        events.push(data);
      },
      onEnd: (_invocationId, data) => {
        ends.push(data);
      },
      onError: (_invocationId, err) => {
        errors.push(err);
      },
    });

    await new Promise((r) => setTimeout(r, 500));
    await handle.cancel();

    const t0 = Date.now();
    await expect(handle.waitEnd()).rejects.toBeInstanceOf(WsCancelledError);
    expect(Date.now() - t0).toBeLessThan(2000);
    expect(ends).toEqual([]);
    expect(errors.length).toBe(1);
    expect(errors[0]).toBeInstanceOf(WsCancelledError);
    expect(events.length).toBeGreaterThanOrEqual(0);
  });
});
